src/text/line.rs

```diff
@@ -586,31 +555,75 @@ impl WidgetRef for Line<'_> {
+                Some(Alignment::Right) => line_width.saturating_sub(area.width),
                 Some(Alignment::Left) | None => 0,
             };
-            x = x.saturating_add(offset);
-            self.to_owned()
+            render_spans(&self.spans, area, buf, offset);
         };
-        for span in &line.spans {
-            let span_width = span.width() as u16;
-            let span_area = Rect {
-                x,
-                width: span_width.min(area.right().saturating_sub(x)),
-                ..area
-            };
-            span.render(span_area, buf);
-            x = x.saturating_add(span_width);
-            if x >= area.right() {
+    }
+}
+
+/// Renders all the spans of the line that are after the `span_offset` in the line and that are
+/// visible in the given area (starting rendering at `area.x + buf_x_offset`). `buf_x_offset`
+/// probably should just be used to modify area.x instead of being a parameter, but it's here
+/// because that would have made the above code just a little uglier
+///
+/// Algorithm:
+/// 1. For each span in the line:
+///    1. If the span is completely before the `span_offset`, skip it
+///    2. If the span is partially before the `span_offset`, render the right side of the span
+///    3. If the span is completely after the `span_offset`, render the whole span
+/// 2. Indent `area` to account for the width of the rendered span
+/// 3. Update `span_offset` to 0 after the first span is rendered in full
+/// 4. Repeat
+fn render_spans(spans: &[Span], mut area: Rect, buf: &mut Buffer, mut span_offset: u16) {
+    for span in spans {
+        let span_width = span.width() as u16;
+        // ignore spans that are completely before the offset
+        if span_offset >= span_width {
+            span_offset -= span_width;
+            continue;
+        }
+        if span_offset > 0 {
+            // the first visible span is only partially visible, so truncate the start and render it
+            // starting at span offset, all subsequent visible spans are rendered in full (as
+            // available space permits)
+            let available_width = span_width - span_offset;
+            let (content, actual_width) = span
+                .content
+                .unicode_truncate_start(available_width as usize);
+            let actual_width = actual_width as u16;
+
+            // if the truncation has truncated an initial grapheme, then we need to start rendering
+            // from a position that takes that into account by indenting the start of the area
+            area = area.indent_x(available_width - actual_width);
+            if area.is_empty() {
+                break;
+            }
+            let right = Span::styled(content, span.style);
+            right.render_ref(area, buf);
+
+            area = area.indent_x(actual_width);
+            span_offset = 0; // ensure that the next span is rendered in full
+        } else {
```

Comment:
Having `span_offset` be `mut` made it hard (for me personally) to understand but I think I get what is happening here now.

Here's some of my comments.

1. the name of the variable is `span_offset` but it really is some kind of width. I think of an `offset` as a point on a line, but here in every calculation you can see the other terms are all "width". fwiw, whenever I read `span_offset` I parsed it as "available_total_span_area_for_rendering".

2. the meat of this function is whether or not to render a partial span. After reading through it, I pictured the code as 3 parts. 
    
a) Skip everything before `span_offset`

b) 
- If the current span is partially renderable, truncate the span
- If the current span is fully renderable, render it

c) Render all the remaining spans

Does it make sense to move the second part into a separate function like `render_maybe_partial_span`?

Also, (I don't know it is possible) I was wondering if it were possible to present the code like so:

```rust
    let mut spans_iter = spans.iter().skip_while(|span| {
       // skip till span_offset
    });

    if let Some(maybe_partial_span) = spans_iter.next() {
       render_maybe_partial_span(...)
    }

    for span in spans_iter {
        // render all the remaining
    }
```

It would have been easier to track the 3 parts for me that way.

All that said, the code + functionality looks good to me overall and happy to merge it to fix this issue.

Reply:
> 1. the name of the variable is `span_offset` but it really is some kind of width. I think of an "offset" as a point on a line, but here in every calculation you can see the other terms are all "width". fwiw, while reading this whenever I came across `span_offset`, I parsed it as `available_total_span_area_for_rendering`.

The offset is the starting point within the line. So `total_span_width - actually_possible_to_display_here = offset`. The offset is definitely not a width. Its the about to be skipped to end up at the width that can be displayed.

---

> Also, (I don't know it is possible) I was wondering if it were possible to present the code like so:

The first iter might be possible relatively easy.

The second and third are kinda hard as the second is involving a condition. Partially or not.

I read the current code kinda like that already. First part to skip everything until we are actually at our offset.

What might simplify this code logic is to reverse the spans and go from right to left. That will be writable, until it isnt. Then abort. The last one might be partial.
But that results in interesting logic to get the correct x. The current code increases the x while the reversed one would reduce it.

I would think of this as an improvement in a later PR. It might be more performance efficient for many spans but I dont know. But we have the tests for many of these cases now so improving this code in the future will be way easier.

Reply:
> The offset is the starting point within the line.

Having the "starting point within the line" change in the algorithm (i.e. be `mut`) made it difficult to read for me. I get that it is "the point that represents the summation of widths that are to be skipped". I guess I was wondering if there was a better way to name it?

Reply:
It's kinda a "width" of the spans to skip... I think `span_skip_width` might make sense as a name.

I can see how offset could a bad name for this, as offset often carries an implication that something is being moved, but here it's referring to how much to move an implicit skip cursor in the spans.

```rust
    let mut spans_iter = spans.iter().skip_while(|span| {
       // skip till span_offset
    });
```
I actually attempted this at one point, but it's difficult to satisfy rust's borrowing rules. To skip the spans, you need to keep a variable that is updated in the closure, but you can't use that variable after the closure as it's moved within the closure. You need that remaining width in order to know the amount to truncate the first span by.

```
error[E0503]: cannot use `span_skip_width` because it was mutably borrowed
   --> src/text/line.rs:606:27
    |
598 |     let spans = spans.iter().skip_while(|span| {
    |                                         ------ `span_skip_width` is borrowed here
599 |         if span_skip_width >= span.width() as u16 {
600 |             span_skip_width -= span.width() as u16;
    |             --------------- borrow occurs due to use of `span_skip_width` in closure
...
606 |     let span_skip_width = dbg!(span_skip_width);
    |                           ^^^^^^^^^^^^^^^^^^^^^ use of borrowed `span_skip_width`
607 |     for span in spans {
    |                 ----- borrow later used here
    |
    = note: this error originates in the macro `dbg` (in Nightly builds, run with -Z macro-backtrace for more info)
```

> The first iter might be possible relatively easy.

Famous last words of a Rust programmer

We could write this first part as a while loop, but it ends up looking pretty much like the for loop. You have to special case the span that you got from `span.next()` when there is span is partially visible, so to do it outside of the loop means either putting the span back into the iterator (i.e. using peekable to not pull each element out, or using [put_back](https://docs.rs/itertools/latest/itertools/fn.put_back.html)), or storing it in a option variable at the end of the loop and checking it. Either way, the added complexity of that code is bad. A better factoring of this logic is to instead do the simplification mentioned in https://github.com/ratatui-org/ratatui/pull/1089#issuecomment-2102340388 (make an method that returns an iterator over the visible spans, including the first partial span).

Rename and comments added in 2fa465c2